python/graphstorm/model/embed.py

```diff
@@ -299,20 +355,38 @@ def forward(self, input_feats, input_nodes):
                 if self.use_node_embeddings:
                     assert ntype in self.sparse_embeds, \
                             f"We need sparse embedding for node type {ntype}"
-                    node_emb = self.sparse_embeds[ntype](input_nodes[ntype], emb.device)
+                    # emb.device: target device to put the gathered results
+                    if is_wholegraph_embedding_module(self.sparse_embeds[ntype]):
+                        # output to local device
+                        node_emb = self.sparse_embeds[ntype](input_nodes[ntype].cuda())
+                        node_emb = node_emb.to(emb.device, non_blocking=True)
+                    else:
+                        node_emb = self.sparse_embeds[ntype](input_nodes[ntype], emb.device)
                     concat_emb=th.cat((emb, node_emb),dim=1)
                     emb = concat_emb @ self.proj_matrix[ntype]
             elif ntype in self.sparse_embeds: # nodes do not have input features
                 # If the number of the input node of a node type is 0,
                 # return an empty tensor with shape (0, emb_size)
                 device = self.proj_matrix[ntype].device
-                if len(input_nodes[ntype]) == 0:
-                    dtype = self.sparse_embeds[ntype].weight.dtype
-                    embs[ntype] = th.zeros((0, self.sparse_embeds[ntype].embedding_dim),
-                                           device=device, dtype=dtype)
-                    continue
-                emb = self.sparse_embeds[ntype](input_nodes[ntype], device)
+                if is_wholegraph_embedding_module(self.sparse_embeds[ntype]):
+                    # Need all procs pass here due to nccl all2lallv in wholegraph
+                    node_emb = self.sparse_embeds[ntype](input_nodes[ntype].cuda())
```

Comment:
What is this? are you forcing every trainer to call wholegraph here?

Reply:
Yes, in case `input_nodes[ntype]` could be an empty tensor and  skip this iteration (`continue` within the following lines); some of the training procs will not call nccl all2all, which is undesirable for wholegraph. Here, I move the wholegraph gather call, i.e., `self.sparse_embeds[ntype](input_nodes[ntype].cuda())` before `continue` to avoid nccl segfault.